### ISSUE TITLE
Support for "Authorization: Bearer" Token Format

### ISSUE THREAD

**Issue by ElmerCat** [nxtbgthng/OAuth2Client#37]
The framework worked correctly for me with SoundCloud, but when I tried WordPress, the response always returned Error 403. It turns out that WordPress requires the token in the HTTP header to be prefixed with "Authorization: Bearer " instead of "Authorization: OAuth ".

I finally got it to work with WordPress by adding a patch to the NXOAuth2Connection.m file, within the createConnection method at this statement:

```
oauthAuthorizationHeader = [NSString stringWithFormat:@"OAuth %@", client.accessToken.accessToken];
```

The patch tests the request URL hostname for the string "wordpress". In that case, it uses "Bearer " instead of "OAuth ":

```
- (NSURLConnection *)createConnection;
   ...
        NSString *host = request.URL.host;
        NSRange range = [host rangeOfString:@"wordpress"];
        if (range.location != NSNotFound) {
            oauthAuthorizationHeader = [NSString stringWithFormat:@"Bearer %@", client.accessToken.accessToken];
        }
        else {
            oauthAuthorizationHeader = [NSString stringWithFormat:@"OAuth %@", client.accessToken.accessToken];
        }
   ...
```

This works for me, but of course it's not a very good way of doing it. Is there some other way to configure the Authorization string format, depending on whatever the different services may require?

**Comment by zuuno** [comment]
Hello,

It's the same for me using "DoorKeeper" oauth provider for Ruby on Rails

**Comment by oliverfoggin** [comment]
Yes, I found this also (this was the issue I was going to open myself)

You can actually see what the type is from the oauth response to the refresh/authorisation request.

In the headers is a header key called "token_type".  The string for this then needs to prefix the access_token when sent back on requests.

I haven't looked properly to find where this can be processed but it will need to be stored against the access_token and used as a parameter for the requests.

Something like...

oauthAuthorizationHeader = [NSString stringWithFormat:@"%@ %@", client.accessToken.tokenType, client.accessToken.accessToken];

Thanks

Oliver

**Comment by oliverfoggin** [comment]
OK, I've written a fix for this.

Will branch it and make a pull request.

Thanks

**Comment by stigi** [comment]
@ojfoggin We love pull requests! Can you branch of the `develop` branch for the changes?

**Comment by oliverfoggin** [comment]
OK, sent, I hope I've done it right.

The change will use the sent token_type.  If the token_type is not received (for whatever reason) it falls back on OAuth.  I had to put a slight fix (for our case-sensitive server) which may not be needed.  The spec says the server should be case-insensitive for token_type so it shouldn't actually matter.

Thanks

Oliver

OK, I hadn't done it right, close pull request and doing again correctly.

OK, resent correctly this time.

**Comment by stigi** [comment]
After reading a bit more about the _bearer_ issue I started wondering if this is the right _time_ to support it.

This framework is build against [draft 10](http://tools.ietf.org/html/draft-ietf-oauth-v2-10) of the OAuth 2 spec, as that was the most current draft at the time we started building it. In the mean time the OAuth 2 situation [has gone bad](http://hueniverse.com/2012/07/oauth-2-0-and-the-road-to-hell/). It's still not final (draft 30 is the most current) & I wonder if the final revision will be any usable at all.

I've got to admit I haven't had a look at any draft after 14, but the [bearer draft](http://tools.ietf.org/html/draft-ietf-oauth-v2-bearer-21) is an extension to OAuth 2 (don't ask me to which draft...).

Because of the entire situation I feel uncomfortable on integrating non draft 10 stuff into the master branch. In an ideal world OAuth2Client would always support the most current draft, or there would be a branch for each. Right I'd rather wait until the spec becomes final and then reevaluate from there rather than running for every new draft.

I hope you see our dilemma & the dilemma of everyone dealing with OAuth2.

For now I'll merge #40 since it's well backwards compatible to draft 10. But I'm afraid of the day when people will need newer drafts which aren't compatible to 10 at all.
